aptos-move/parallel-executor/src/executor.rs

```diff
@@ -325,22 +350,33 @@ where
 
         // Extract outputs in parallel.
         let num_txns = scheduler.num_txn_to_execute();
-        let valid_results_size = AtomicUsize::new(num_txns);
+        let results_info = Mutex::new(ExecutionStatusInfo::<E>::new(num_txns));
         let chunk_size = (num_txns + 4 * self.concurrency_level - 1) / (4 * self.concurrency_level);
-        RAYON_EXEC_POOL.install(|| {
+        let final_results = RAYON_EXEC_POOL.install(|| {
             (0..num_txns)
                 .collect::<Vec<TxnIndex>>()
                 .par_chunks(chunk_size)
-                .map(|chunk| {
+                .flat_map(|chunk| {
+                    let mut chunk_results = Vec::with_capacity(chunk_size);
                     for idx in chunk.iter() {
                         let res = last_input_output.take_output(*idx);
-                        if matches!(res, ExecutionStatus::SkipRest(_)) {
-                            valid_results_size.fetch_min(*idx + 1, Ordering::SeqCst);
-                        }
-                        outcomes.set_result(*idx, res);
+
+                        match res {
+                            ExecutionStatus::Success(t) => {
+                                chunk_results.push(t);
+                            }
+                            ExecutionStatus::SkipRest(t) => {
+                                chunk_results.push(t);
+                                results_info.lock().process_skip_rest(*idx);
```

Comment:
I found the lock a bit ugly, how much do we get from this par_chunk? can't we just do one pass iteration over last_input_output like the one we did in outcome_arrary?